### ISSUE TITLE
Bug Report: Correct the implementation of Pagination across Talawa-Admin.

### ISSUE THREAD

**Issue by beingnoble03** [PalisadoesFoundation/talawa-admin#676]
**Describe the bug**
We have proper queries in the `talawa-api` that can provide pagination but still these are not used in `talawa-admin`.

**Expected behavior**
A proper and correct implementation of pagination. By resending the query for every page, when it is changed, and fetching only the data that has to be displayed in that page.

**Actual behavior**
Improper implementation of pagination by bringing all the data at once, and handling the data in the frontend.

**Screenshots**

**Additional details**

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

**Comment by beingnoble03** [comment]
@palisadoes please assign the issue to me. Thanks!

**Comment by beingnoble03** [comment]
Also, as this is a functionality based issue, the code can be reused even after UI revamp.

**Comment by beingnoble03** [comment]
@palisadoes this requires some work in talawa-api. Will open an issue for the same.

**Comment by beingnoble03** [comment]
@palisadoes @xoldyckk This seems like a big issue now.
To implement pagination in `talawa-admin`, `organizationsConnection` has to be implemented correctly in `talawa-api`, as it doesn't return the pagesCount, which is necessary for page-based pagination. So, it needs to be refactored, but if it is implemented correctly, as this query `organizationsConnection` is used in `talawa` too. So, we will need to make required changes there too. Should I proceed by creating issues in `talawa-api` and `talawa`?

**Comment by xoldyckk** [comment]
@beingnoble03 why is page based pagination needed?

**Comment by beingnoble03** [comment]
@xoldyckk all paginations are page based in `talawa-admin`. But, yes, I agree, we don't need it to be page based. Should we change all the paginators in `talawa-admin` so that in future it would be easy to migrate when redesigning the UI?

**Comment by beingnoble03** [comment]
@palisadoes please look at this. Should we replace page based pagination with infinite scroll or just next-previous without showing the total count?

**Comment by beingnoble03** [comment]
@palisadoes For posts, the `talawa-api` was already configured for pagination. So, I implemented pagination in posts.
Should I change it to infinite scrolling?
The problem lies on `organizationsConnection` query. To implement correct pagination for it, I need to open corresponding issue in `talawa-api`.

**Comment by palisadoes** [comment]
1. Please create the corresponding issue. 
2. The posts functionality needs to be like that of any other app that has a news feed.

**Comment by xoldyckk** [comment]
@palisadoes

I think you're mixing up what pagination means. Infinite scrolling is a kind of pagination too. You have two main kinds of pagination offset and cursor based.

Offset based(also called page based) pagination is where you literally go line by line skipping them till you skip the number of rows specified in offset. So, offset here is basically `number of pages you want to skip * number of items on a page`. The more amount of rows the worse this performs, because database engine traverses the table going through each row one by one instead of directly jumping to the desired row. The notion of pages like `1 2 3 4 5` is exactly this, the more number of pages you have the worse database will perform when you try to access the pages with bigger values. In this approach data on the pages is not consistent, the data can be added/removed therefore leading to shifting/deleting of items which previously existed on a particular page. This kind of pagination was prevalent in old times, but it's not the recommended approach anymore.

Cursor based pagination is where you use primary/unique indexed columns of rows. Traversal using these columns are fast because databases are fast at traversing rows using indexed columns. Cursor is the primary/unique indexed column you want the data before or after. The notion of pages like `1 2 3 4 5` etc., doesn't exist here because with one cursor you can only go forward and backward in respect to position of the row that cursor corresponds to. So, no jumping to pages here.

But jumping to pages is not something so grandiose. Instead of this simply sorting and filtering capabilities can be provided to clients to get to the results they want. So, for something like organizations page you can simply provide the ability to only show organizations starting with a particular alphabet/word etc. Infinite scrolling can also be implemented using offset based pagination, it's not about that, it's about the performance cost you incur when using offset based pagination.

Graphql connections are the standard for pagination in graphql ecosystem as they're specified in graphql server specification and they make use of cursor based pagination. So, to implement graphql connections you have to implement cursor pagination. I'd recommend sticking to standards.

The only possible benefit I see for page based pagination is that users will be able to simply jump to pages instead of scrolling or manually providing filters to get the results they want. But I'd also implore you to consider the things I explained above. Let me know your rationale for why page based pagination is needed.

**Comment by beingnoble03** [comment]
@palisadoes @xoldyckk Please this needs to be cleared as soon as possible.
Also, If we are changing the implementation in `talawa-api`, then this would mean that we would also need to implement pagination in `talawa` (mobile-app) correctly too. For example, `organizationsConnection` is being used in `talawa-mobile-app` too.

**Comment by palisadoes** [comment]
Use what Graphql excels at. Cursor based would therefore be the best approach.

The search feature is not going to return infinite results. It will be really limited. Tags will limit the results even further.

**Comment by prathamesh-mutkure** [comment]
Shopify has a good guide on GraphQL design patterns which includes pagination

This might be helpful to this discussion - https://github.com/Shopify/graphql-design-tutorial/blob/master/TUTORIAL.md#lists-and-pagination

**Comment by beingnoble03** [comment]
@xoldyckk So, we need to implement a pagination in `talawa-admin` which doesn't show total number of pages. Right?
I will open issues required on `talawa` and `talawa-api` too.

**Comment by beingnoble03** [comment]
I searched on `talawa`, found that only `organizationsConnection` is the only Connection being used there.

**Comment by xoldyckk** [comment]
@beingnoble03 At the current moment there isn't a single resolver in talawa-api which implements graphql connection. It's just named `organizationsConnection`, but it isn't actually a connection. Many lists in talawa-api are returned as `[Item]` type. So, the whole list is returned in one go, there's no pagination.

**Comment by palisadoes** [comment]
@beingnoble03 The answer is implement cursor based pagination in all cases unless by extreme exception. It'll also mean that we'll need to properly implement GraphQL connections to achieve this.

**Comment by beingnoble03** [comment]
@xoldyckk Yaa, I've already seen that. I was just saying that `organizationConnection` is used in talawa-mobile-app, So, I need to make required changes in the `talawa` repo.

@palisadoes Yeah. Working on it.

**Comment by palisadoes** [comment]
We could do something like this in some cases for pagination. 

1. As the person clicks `next` the box to the right get incremented.
2. If the person wants to go to a specific page, they just type in the number. 

Is tracking this possible with cursors? It would improve the user experience.

![image](https://user-images.githubusercontent.com/16875803/229223771-42b4e001-ba7b-49d5-aa29-0715b946d838.png)

**Comment by palisadoes** [comment]
But it could be tracked on the client side where the node ID is always returned in the query results, then the `first` and `after` query clauses could be used to implement this. Correct?

**Comment by beingnoble03** [comment]
@palisadoes Actually, we need 2 parameters `after` which is the ID of the last object on the previous page (for the page we want to jump into) and `first` which tells the query resolver how many objects do we need to return.
So, as we would be missing the `after` paramter, this isn't possible.

**Comment by beingnoble03** [comment]
Sorry, for the delayed response. I've been travelling for 2 days.

**Comment by palisadoes** [comment]
Wouldn't it be possible if the API always responded with the ID values of the nodes?

**Comment by beingnoble03** [comment]
@palisadoes @xoldyckk I was reading about cursor pagination from this [blog](https://uxdesign.cc/why-facebook-says-cursor-pagination-is-the-greatest-d6b98d86b6c0).

We have implemented sorting along various columns in `talawa-api`. Now, for implementing cursor based pagination, we need to limit these sorting parameters to only unique columns, else we might loose benefit of using cursor based pagination, which will definitely affect the performance.

In this [another blog](https://medium.com/swlh/how-to-implement-cursor-pagination-like-a-pro-513140b65f32), author says that if we want to sort on the basis of a non-unique column, we will need to create another unique column by concatenating 2 or more columns. Which means we are creating a column just for sorting.

> If you don’t want to add a new column to your database just for sorting purposes, consider abandoning cursor pagination entirely and instead use the default offset pagination instead as it will perform better for earlier pages, which users will encounter most often.

A statement from the blog.


So, we are left with 2 options here,
- We limit the sorting feature.
- Using offset pagination for the queries having non-unique column sort functionality.

Please do let me know, which option do we want.

**Comment by beingnoble03** [comment]
> Wouldn't it be possible if the API always responded with the ID values of the nodes?

@palisadoes Cursor pagination was made for infinite scrolling. So, it is not possible to get to a random page using cursor based pagination using any method, because we need the endCursor of the previous page for fetching the items of the next page. This is one of the disadvantages of the cursor based pagination.

**Comment by beingnoble03** [comment]
We need to finalize this before GSoC period. Can be an obstruction for many UI related ideas.

**Comment by palisadoes** [comment]
What are the most likely fields we would need to sort against and how could we best ensure uniqueness?

**Comment by palisadoes** [comment]
I think we can easily enforce organization name uniqueness, but people names will be difficult.

So for people, we may need a sortable unique column.

Events would be sortable by date time which may not be unique if two events start at the same time, so a solution for that is needed too.

Posts would be sorted by date time. But it could be solved by storing time to the millisecond.

What other exceptions can you think of that may require this?

**Comment by xoldyckk** [comment]
> I think we can easily enforce organization name uniqueness, but people names will be difficult.
> 
> So for people, we may need a sortable unique column.
> 
> Events would be sortable by date time which may not be unique if two events start at the same time, so a solution for that is needed too.

Yes, this can be done, but the user will only be able to go back and forth between consecutive pages, not random pages, like they can't go from page 1 to page 4 directly, they'll have to go through page 2 and page 3 too.

I have explained all about this over [here](https://github.com/PalisadoesFoundation/talawa-admin/issues/676#issuecomment-1487264419).

You can use any kind of sort and filter on any field you want, but on large datasets they'll perform worse without indexing and the complexity of filter condition.

**Comment by palisadoes** [comment]
The undesirability of indexing for random pages is understood and accepted. I don't think we'll need it if we have the correct filters and the identification of ways to create unique keys as I was explaining before.

Are these assumptions correct?

**Comment by xoldyckk** [comment]
> I think we can easily enforce organization name uniqueness, but people names will be difficult.
> 
> So for people, we may need a sortable unique column.
> 
> Events would be sortable by date time which may not be unique if two events start at the same time, so a solution for that is needed too.
> 
> Posts would be sorted by date time. But it could be solved by storing time to the millisecond.
> 
> What other exceptions can you think of that may require this?

Fields don't need to be unique to be able to be filtered or sorted by. To retrieve results for these kinds of filters/sorts you still need to provide an primary id or a unique column to identify where you left off after returning the list of items first time they were asked for while paginating. 

People might have same name so sorting might sound silly here. In such situtations you can only sort them by their `createdAt` field, their primary ids or some other unique column etc. So, first you apply the conditions for filter/sort using their name and if many people are encountered with the same name, you sort them by the fields I told about.

Basically any exception you can think of is solvable. When fields are not unique, you pick a generic thing like `createdAt` or `id` field to sort them.

**Comment by palisadoes** [comment]
Let's proceed as you both have advised.

**Comment by beingnoble03** [comment]
> > I think we can easily enforce organization name uniqueness, but people names will be difficult.
> > So for people, we may need a sortable unique column.
> > Events would be sortable by date time which may not be unique if two events start at the same time, so a solution for that is needed too.
> > Posts would be sorted by date time. But it could be solved by storing time to the millisecond.
> > What other exceptions can you think of that may require this?
> 
> Fields don't need to be unique to be able to be filtered or sorted by. To retrieve results for these kinds of filters/sorts you still need to provide an primary id or a unique column to identify where you left off after returning the list of items first time they were asked for while paginating.
> 
> People might have same name so sorting might sound silly here. In such situtations you can only sort them by their `createdAt` field, their primary ids or some other unique column etc. So, first you apply the conditions for filter/sort using their name and if many people are encountered with the same name, you sort them by the fields I told about.
> 
> Basically any exception you can think of is solvable. When fields are not unique, you pick a generic thing like `createdAt` or `id` field to sort them. This isn't limited to complete name, for example take something like all organizations that have names starting with letter `A`, here you'd first sort by the remaining words in their names that come after the letter `A`, if the complete names of organizations are equal as well, then you opt for fields like `createdAt` or `id`.

- If the field isn't unique, we will basically, sort the results using 2 columns, i.e. `name` and `id`. So, are we supposed the pass the `cursor` as `name_id`, when we are selecting "Sort By Name" on the client side?

- But, this will downgrade the performance by a significant amount. The time complexity will be worse than offset based pagination as we are sorting the whole data using 2 fields (one of them being unique). Should we create an index for `name` and `id`? [i.e creating an index for `non-unique` field and `unique` field]

**Comment by beingnoble03** [comment]
@xoldyckk @palisadoes 

Also, we need to limit the options for sorting. Currently, we have option for sorting for almost every field.
For Events, we can have
- `title`
- `startDate`
- `allDay`
- `id`

For Organization, we can have
- `name`
- `id`

For People, we can have
- `firstName`
- `email`
- `id`

For Post, we can have
- `commentCount`
- `createdAt`
- `likeCount`
- `title`
- `id`

Are there any other fields we need these to be sorted?

**Comment by xoldyckk** [comment]
> > > I think we can easily enforce organization name uniqueness, but people names will be difficult.
> > > So for people, we may need a sortable unique column.
> > > Events would be sortable by date time which may not be unique if two events start at the same time, so a solution for that is needed too.
> > > Posts would be sorted by date time. But it could be solved by storing time to the millisecond.
> > > What other exceptions can you think of that may require this?
> > 
> > 
> > Fields don't need to be unique to be able to be filtered or sorted by. To retrieve results for these kinds of filters/sorts you still need to provide an primary id or a unique column to identify where you left off after returning the list of items first time they were asked for while paginating.
> > People might have same name so sorting might sound silly here. In such situtations you can only sort them by their `createdAt` field, their primary ids or some other unique column etc. So, first you apply the conditions for filter/sort using their name and if many people are encountered with the same name, you sort them by the fields I told about.
> > Basically any exception you can think of is solvable. When fields are not unique, you pick a generic thing like `createdAt` or `id` field to sort them. This isn't limited to complete name, for example take something like all organizations that have names starting with letter `A`, here you'd first sort by the remaining words in their names that come after the letter `A`, if the complete names of organizations are equal as well, then you opt for fields like `createdAt` or `id`.
> 
> * If the field isn't unique, we will basically, sort the results using 2 columns, i.e. `name` and `id`. So, are we supposed the pass the `cursor` as `name_id`, when we are selecting "Sort By Name" on the client side?
> * But, this will downgrade the performance by a significant amount. The time complexity will be worse than offset based pagination as we are sorting the whole data using 2 fields (one of them being unique). I am sure, we will need to create an index for `name` and `id` [i.e creating an index for `non-unique` field and `unique` field]
> 
> A compound index will be created `{ name: 1, _id: 1 }`.

Whether you pass the name in cursor as well is up to you, it's not necessary, cursor should be something that uniquely identifies that edge. The context of edge can be different in different situations. You encode in cursor information which uniquely identifies that edge. For example take a `post`. There's popular posts, trending posts, recommended posts etc., so many different kinds of connections all related to post but they mean different things. A post might be present in all of these connections but the post edge itself is related  to different connections, so the cursor here will reference the post edge for that particular connection. All these are very high level concepts and I've never seen the use of cursors in advanced ways. 

For most of the implementations cursor can be the same thing as postId, though not the other way around. 

Idk what you meant by time complexity and why offset based pagination will be better, the whole point of cursor based pagination is to directly jump to the object that cursor represents, that's all. The cost you incur filtering/sorting with many fields has nothing to do with it.

**Comment by beingnoble03** [comment]
@xoldyckk What I meant was with time complexity thing was that sorting the data by 2 columns will drop the performance with significant amount. But, yes, obviously this can avoided if we create a compound index having 2 fields:
- field which we need to sort that data
- `_id`.

**Comment by xoldyckk** [comment]
Yes, but that has nothing to do with cursor or offset based pagination.

**Comment by xoldyckk** [comment]
If you want faster filtering/sorting you will have to create indexes

**Comment by beingnoble03** [comment]
I think, in offset based pagination, we would be sorting by just one column, and then paginating, whereas in cursor based, we would have to sort the data with 2 columns [in the case of sorting with non-unique column]

<img width="710" alt="image" src="https://user-images.githubusercontent.com/62551163/230717309-ce594b68-98d5-4674-9995-6dc320bbff65.png">

This table is what I am talking about.

**Comment by beingnoble03** [comment]
> If you want faster filtering/sorting you will have to create indexes, creating indexes also increases the resource consumption of database. It's a tradeoff, you make tradeoffs considering the business requirements of your application.
> 
> If the client is never going to use a filter functionality you're wasting the database resources indexing that filter column.

Yes, that's why I am confirming the columns we want to be sorted.

**Comment by beingnoble03** [comment]
> @xoldyckk @palisadoes
> 
> Also, we need to limit the options for sorting. Currently, we have option for sorting for almost every field. For Events, we can have
> 
> * `title`
> * `startDate`
> * `allDay`
> * `id`
> 
> For Organization, we can have
> 
> * `name`
> * `id`
> 
> For People, we can have
> 
> * `firstName`
> * `email`
> * `id`
> 
> For Post, we can have
> 
> * `commentCount`
> * `createdAt`
> * `likeCount`
> * `title`
> * `id`
> 
> Are there any other fields we need these to be sorted?

Can you confirm this once?

**Comment by palisadoes** [comment]
People will also need to be sorted by last, first names.

People won't need to sort by post "like" and comment counts. Chronology is sufficient.

AllDay would be for filtering, wouldn't it?

I've never had to sort by email address that I can remember. People use so many random names that it would be meaningless.

Other than that, is OK

**Comment by beingnoble03** [comment]
- I guess for People, last and first names can be indexed and sorted together.
- Hmm, I guess AllDay can be removed from sorting.
- Sorting by email can be kept as it will not require any extra work. I am assuming `email` to be unique for every user.

**Comment by palisadoes** [comment]
Email addresses won't necessarily be unique. Spouses will share email addresses for various reasons.

**Comment by palisadoes** [comment]
However for the purposes of a unique login, week need to assume uniqueness.
